### PR TITLE
lgz/training_program

### PR DESCRIPTION
**Pull request by a8520238**
1th training_program

### CODE REVIEW

**Review by dutlzn — DISMISSED** [review]

**Review by ghh82058 — APPROVED** [review]

**Review by dutlzn — APPROVED** [review]

File: training_program/urls.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+'''Register URL routes in auth module.'''
+from rest_framework import routers
+import training_program.views
+
+router = routers.SimpleRouter()
+router.register(r'programcategory', training_program.views.
```
Comment (ghh82058) [inline training_program/urls.py]:
```suggestion
router.register(r'program-categories', training_program.views.
```

File: training_program/urls.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+                ProgramCategoryViewSet)
+router.register(r'programform', training_program.views.ProgramFormViewSet)
```
Comment (ghh82058) [inline training_program/urls.py]:
```suggestion
router.register(r'program-forms', training_program.views.ProgramFormViewSet)
```

**Review by ghh82058 — COMMENTED** [review]
A few changes.

File: TMSFTT/urls.py [hunk 3]
```diff
@@ -21,6 +21,7 @@
 API_URLPATTERNS = [
     path('auth/', include('auth.urls')),
     path('training_event/', include('training_event.urls')),
+    path('training_program/', include('training_program.urls')),
```
Comment (Time1ess) [inline TMSFTT/urls.py]:
```suggestion
    path('training-program/', include('training_program.urls')),
```
I think dash is a better practice than underscore, and sorry for the bad example, I've fixed it.

File: training_program/tests/tests_serializers.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+import training_program.serializers as serializers
+import training_program.models as models
+
+
+class TestProgramCategorySerializer(TestCase):
+    '''Unit tests for serializer of programcategory.'''
+    def test_fields_equal(self):
+        '''Serializer should return fields of ProgramCategory correctly.'''
+        programcategory = models.ProgramCategory()
```
Comment (Time1ess) [inline training_program/tests/tests_serializers.py]:
```suggestion
        category = models.ProgramCategory()
```
Given the context, It's ok to name the variable as `category`.

File: training_program/tests/tests_serializers.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        expected_keys = {'id', 'name'}
+        keys = set(serializers.ProgramCategorySerializer
+                   (programcategory).data.keys())
```
Comment (Time1ess) [inline training_program/tests/tests_serializers.py]:
```suggestion
                   (category).data.keys())
```
ditto.

File: training_program/tests/tests_serializers.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        self.assertEqual(keys, expected_keys)
+
+
+class TestProgramFormSerializer(TestCase):
+    '''Unit tests for serizalizer of programform.'''
+    def test_fields_equal(self):
+        '''Serializer should return fields of ProgramForm correctly.'''
+        programform = models.ProgramForm()
```
Comment (Time1ess) [inline training_program/tests/tests_serializers.py]:
```suggestion
        form = models.ProgramForm()
```
Ditto.

File: training_program/tests/tests_serializers.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        expected_keys = {'id', 'name'}
+        keys = set(serializers.ProgramFormSerializer(programform).data.keys())
```
Comment (Time1ess) [inline training_program/tests/tests_serializers.py]:
```suggestion
        keys = set(serializers.ProgramFormSerializer(form).data.keys())
```
Ditto.

File: training_program/tests/tests_serializers.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        self.assertEqual(keys, expected_keys)
+
+
+class TestProgramSerializer(TestCase):
+    '''Unit tests for serizalizer of program.'''
+    def test_fields_equal(self):
+        '''Serializer should return fields of Program correctly.'''
+        program = models.Program()
+        expected_keys = models.Program()
```
Comment (Time1ess) [inline training_program/tests/tests_serializers.py]:
```suggestion
```
Unused?

File: training_program/tests/tests_views.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    def test_create_programcategory(self):
+        '''ProgramCategory should be created by POST request.'''
+        url = reverse('programcategory-list')
+        name = 'programcategory'
+        data = {'name': name}
+
+        response = self.client.post(url, data, format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(training_program.models.ProgramCategory.
```
Comment (Time1ess) [inline training_program/tests/tests_views.py]: https://github.com/DLUT-SIE/TMSFTT-BE/blob/24e4155fe4aa1002811d9fefe8a34f04192526e2/training_program/tests/tests_views.py#L22-L23
You can use:
```
self.assertEqual(
    training_program.models.ProgramCategory.objects.count(), 1)
```
or use a parenthesis to wrap your chain calls, like:
```
self.assertEqual((training_program.models.ProgramCategory
                 .objects.count()), 1)
```
Note that we move the dot to the next line and add a `()` to wrap the whole chain call.

IMHO, the former is a better practice, unless your code is too long.

File: training_program/tests/tests_views.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+    def test_delete_programcategory(self):
+        '''Programcategory list should be deleted by DELETE request.'''
+        programcategory = mommy.make(training_program.models.ProgramCategory)
+        url = reverse('programcategory-detail', args=(programcategory.pk,))
+
+        response = self.client.delete(url)
+
+        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
+        self.assertEqual(training_program.models.ProgramCategory.objects.
```
Comment (Time1ess) [inline training_program/tests/tests_views.py]: Ditto.

File: training_program/tests/tests_views.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+        url = reverse('programform-list')
+        name = 'programform'
+        data = {'name': name}
+
+        response = self.client.post(url, data, format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(training_program.models.ProgramForm.objects.count(),
+                         1)
+        self.assertEqual(training_program.models.ProgramForm.objects.get().
```
Comment (Time1ess) [inline training_program/tests/tests_views.py]: Ditto.

File: training_program/tests/tests_views.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+from rest_framework import status
+from rest_framework.test import APITestCase
+
+import training_program.models
+import auth.models
+
+
+class TestProgramCategoryViewSet(APITestCase):
+    '''Unit tests for ProgramCategory view.'''
+    def test_create_programcategory(self):
```
Comment (Time1ess) [inline training_program/tests/tests_views.py]:
```suggestion
    def test_create_program_category(self):
```

File: training_program/tests/tests_views.py [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+
+        response = self.client.post(url, data, format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(training_program.models.ProgramCategory.
+                         objects.count(), 1)
+        self.assertEqual(training_program.models.ProgramCategory.objects.
+                         get().name, name)
+
+    def test_list_programcatgegery(self):
```
Comment (Time1ess) [inline training_program/tests/tests_views.py]:
```suggestion
    def test_list_program_catgegery(self):
```

File: training_program/tests/tests_views.py [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+        '''Programcategory list should be accessed by GET request.'''
+        url = reverse('programcategory-list')
+
+        reponse = self.client.get(url)
+
+        self.assertEqual(reponse.status_code, status.HTTP_200_OK)
+
+    def test_delete_programcategory(self):
```
Comment (Time1ess) [inline training_program/tests/tests_views.py]:
```suggestion
    def test_delete_program_category(self):
```

File: training_program/tests/tests_views.py [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+        url = reverse('programcategory-list')
+        name = 'programcategory'
+        data = {'name': name}
+
+        response = self.client.post(url, data, format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(training_program.models.ProgramCategory.
+                         objects.count(), 1)
+        self.assertEqual(training_program.models.ProgramCategory.objects.
```
Comment (Time1ess) [inline training_program/tests/tests_views.py]: Ditto.

File: training_program/tests/tests_views.py [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+        '''Programcategory list should be accessed by GET request.'''
+        url = reverse('programcategory-list')
+
+        reponse = self.client.get(url)
+
+        self.assertEqual(reponse.status_code, status.HTTP_200_OK)
+
+    def test_delete_programcategory(self):
+        '''Programcategory list should be deleted by DELETE request.'''
+        programcategory = mommy.make(training_program.models.ProgramCategory)
```
Comment (Time1ess) [inline training_program/tests/tests_views.py]:
```suggestion
        category = mommy.make(training_program.models.ProgramCategory)
```
Ditto.

File: training_program/tests/tests_views.py [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+
+        response = self.client.delete(url)
+
+        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
+        self.assertEqual(training_program.models.ProgramCategory.objects.
+                         count(), 0)
+
+    def test_get_programcategory(self):
+        '''Programcategory list should be GET by GET request.'''
+        programcategory = mommy.make(training_program.models.ProgramCategory)
```
Comment (Time1ess) [inline training_program/tests/tests_views.py]:
```suggestion
        category = mommy.make(training_program.models.ProgramCategory)
```
Ditto.

File: training_program/tests/tests_views.py [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+        response = self.client.get(url)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(set(response.data.keys()), expected_keys)
+
+    def test_update_programcategory(self):
+        '''Programcategory list should be updated by PATCH request.'''
+        name0 = 'programcategory0'
+        name1 = 'programcategory1'
+        programcategory = mommy.make(training_program.models.ProgramCategory,
```
Comment (Time1ess) [inline training_program/tests/tests_views.py]:
```suggestion
        category = mommy.make(training_program.models.ProgramCategory,
```
Ditto.

File: training_program/tests/tests_views.py [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+        '''ProgramForm list should be accessed by GET request.'''
+        url = reverse('programform-list')
+
+        response = self.client.get(url)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_delete_programform(self):
+        '''ProgramForm list should be deleted by DELETE request.'''
+        programform = mommy.make(training_program.models.ProgramForm)
```
Comment (Time1ess) [inline training_program/tests/tests_views.py]:
```suggestion
        form = mommy.make(training_program.models.ProgramForm)
```
Ditto.

File: training_program/tests/tests_views.py [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+        url = reverse('programform-detail', args=(programform.pk,))
+
+        response = self.client.delete(url)
+
+        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
+        self.assertEqual(training_program.models.ProgramForm.objects.count(),
```
Comment (Time1ess) [inline training_program/tests/tests_views.py:105]: Ditto.

File: training_program/tests/tests_views.py [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+                         0)
+
+    def test_get_programform(self):
+        '''ProgramForm list should be GET by GET request.'''
+        programform = mommy.make(training_program.models.ProgramForm)
```
Comment (Time1ess) [inline training_program/tests/tests_views.py]: Ditto.

File: training_program/tests/tests_views.py [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+        response = self.client.get(url)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(set(response.data.keys()), expected_keys)
+
+    def test_update_programform(self):
+        '''ProgramForm list should be updated by PATCH request.'''
+        name0 = 'programForm0'
+        name1 = 'programForm1'
+        programform = mommy.make(training_program.models.ProgramForm,
```
Comment (Time1ess) [inline training_program/tests/tests_views.py]: Ditto.

File: training_program/tests/tests_views.py [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+        programform = mommy.make(training_program.models.ProgramForm)
+        url = reverse('programform-detail', args=(programform.pk,))
+        expected_keys = {'id', 'name'}
+
+        response = self.client.get(url)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(set(response.data.keys()), expected_keys)
+
+    def test_update_programform(self):
```
Comment (Time1ess) [inline training_program/tests/tests_views.py]: Ditto.

File: training_program/urls.py (hunk 1 above)
Comment (Time1ess) [reply training_program/urls.py]: +1

File: training_program/urls.py (hunk 2 above)
Comment (Time1ess) [reply training_program/urls.py]: +1

**Review by Time1ess — CHANGES_REQUESTED** [review]
* Variable naming style, like `programcategory` could be `category` given a context.
* Function naming style, like `test_create_programcategory` should be `test_create_program_category`.
* How do we indent codes in Python ? You can refer to [link](https://www.python.org/dev/peps/pep-0008/#indentation).
* Routing naming style. In practice, we use `-`(dash) more, instead of `_`(underscore).

**Review by Time1ess — APPROVED** [review]

**Pull request merged**